### ISSUE TITLE
Appears that variables outside of macro blocks are not in scope inside the macro?

### ISSUE THREAD

**Issue by wesruv** [twigphp/Twig#2171]
Correct me if I'm wrong, but that seems to be the case. So:

```
{% set foo = 'wakka' %}

{% macro bar(thing1) %}
  {{ foo }}
  {# ^ Doesn't print anything #}
{% endmacro %}
```

Would be good to put a note about that on the `macro` page and the `set` page. If there's any other elements that define their own scope it'd be good to have a list on the set page.

**Comment by stof** [comment]
macros are fully scoped. They only have access to their arguments (and global variables). This behavior is logical, as they are meant to be imported into other templates and so would not have any idea about what variables are there (and the ``{% set foo = 'wakka' %}`` of the variable defining the macro would never be executed when using the macro in another template). Think about macros in a way similar to PHP functions (which is what they are under the hood btw).

The other scoping element is ``{% block %}``. Each block starts a new scope (initializing it with the current context), which means that any ``{% set %}`` tag inside a block does not affect the context outside the block

**Comment by joelpittet** [comment]
Seems reasonable to add it to the macro page so people don't assume JS like scope.

**Comment by apfelbox** [comment]
@joelpittet see

> But as with PHP functions, macros don't have access to the current template variables.

http://twig.sensiolabs.org/doc/tags/macro.html

**Comment by wesruv** [comment]
Ahh, I missed that. FWIW, I'm not sure I would have taken it's meaning if I had... but maybe that's a 'me' problem.

Thanks!

**Comment by wesruv** [comment]
> The other scoping element is {% block %}. Each block starts a new scope (initializing it with the current context), which means that any {% set %} tag inside a block does not affect the context outside the block

Ah, this makes sense! Are there any docs on variable scope? I think I'd find that helpful as PHP is not one of my major skillsets, I write enough PHP to spit out the HTML/CSS I want.

Anyway, just a thought, thanks again for the clarifications!

**Comment by fabpot** [comment]
@wesruv It would be very useful for us if you could tell us how to change the sentence to make it more understandable for you. It's always great to get feedback from someone who does not necessarily have a PHP background.

I'm going to reopen this issue as having a section on scopes would be a great addition to the docs. Right now, scopes are defined in different places but having a dedicated paragraph makes sense.

**Comment by wesruv** [comment]
> But as with PHP functions, macros don't have access to the current template variables.

Maybe... 

> But macros do not have access to the variables that were set outside of them, just like PHP functions.

If there's a specific page or section regarding scope it could be linked to after this bit which would help with 'scannability'.

I would like to say that the Twig docs are _awesome_, I've used them a lot (on Drupal 8) and have been able to learn a lot quite easily and quickly. So kudos! 😄 

In case understanding my specific use-case helps:
When I was trying to understand the problem I was running into I did a 'Find' for 'scope'  on `set` and `macro` pages and didn't come up with anything. I was in 'debug' mode and not 'learning' mode, I wasn't even sure if scope was the issue and didn't read the docs thoroughly.

Before I made this issue I read through macro and set more, but didn't catch the part quoted above.

Cheers!
